Fix/abi spec v0.6.0 breaking change

xMNG opened the pull request:
This PR fixes #74 . 

These changes allow drizzle to properly distinguish between functions that need to be bound to `cacheCall` vs `cacheSend`. Also included are unit tests and a separate UI test (see test-0.6.0-ABI under ui-tests) to verify that drizzle can consume 0.6.0+ ABI artifacts while preserving backwards compatibility with 0.5.0+ artifacts. The UI test uses a solc 0.6.0+ contract that incorporates both `payable` and `nonpayable` functions.

To run the unit tests: 
1. In root dir: `yarn`
2. Run unit tests: `yarn test`

To run the UI test:
1. In root dir: `yarn`
2. In another terminal, deploy a local blockchain: `yarn ganache` 
3. Change dir to the truffleV6 project: `cd ui-tests/truffleV6`
4. Compile and migrate the 0.6.0 contract: `yarn ganache:deploy`
5.  Change dir to ui test: `../test-0.6.0-ABI`
6. Launch the ui test `yarn serve:ui`


CC: @nicholaspai -  please review and let us know if there are any changes or missing items. Thank you!

Co-authored by @MalloryLeeWong 
Thanks @cds-amal for your feedback and guidance.

josecelano commented:
Is there any reason why this fix is not merged yet?

JFGHT commented:
Things are going a bit slow in here, just like this important issue:
https://github.com/trufflesuite/drizzle/issues/59

which hasn't been fixed either.

Commit:
fix: issue #74

Solidity 0.6 introduced a new representation of `constant` and `payable`
functions that broke drizzle's implementation to identify and categorize
functions as `cacheCall` and 'cacheSend'

  - change implementation to identify constant functions for all current
    versions of Solidity in order to be backward compatible.
  - refactor functions to test constants
  - add unit tests for refactored functions

MalloryLeeWong committed:
test: create React app for integration testing

React app is needed to test Truffle projects using solc >=0.6.0.

Commit:
test: add trufflev6 files in order to test 0.6.0+ compatibility

This Truffle project is bound to a solc 0.6.0+ and will allow us to test
that drizzle can consume the updated ABI spec properly.

Commit:
test: update test ui with descriptions

Commit:
build: update gitignore to ignore build artifacts

Commit:
test: handle arrays

Update contract to return arrays and update UI to properly display
arrays

nicholaspai commented on packages/store/src/DrizzleContract.js:
```diff
@@ -22,20 +22,18 @@ class DrizzleContract {
       var item = this.abi[i]
-
-      if (item.type == 'function' && item.constant === true) {
+      if (isGetterFunction(item)) {
```
+1 for moving this logic to its own module

cds-amal commented on ui-tests/truffle/migrations/2_deploy_contracts.js:
```diff
@@ -1,9 +1,11 @@
 const SimpleStorage = artifacts.require("SimpleStorage");
 const TutorialToken = artifacts.require("TutorialToken");
 const ComplexStorage = artifacts.require("ComplexStorage");
+const Grue = artifacts.require("Grue");
```
Are these contracts used?

Commit:
test: remove Grue contract and migration

Commit:
test: remove duplicate simpleStorage test
